### ISSUE TITLE
mongodb ->with() foreach relation

### ISSUE THREAD

**Issue by bdart** [yiisoft/yii2#8165]
Discuss as MySQL model has an relationship to a mongo Model with this:
```php
    public function getDiscussContent()
    {
        return $this->hasOne(MongoDiscussion::className(), ['_id'=>'storage_id']);
    }
```
If I pull my results with:
```php
        $table = Discuss::find()
            ->where(['refer_id' => $id,'refer_type' => 'article'])
            ->with('discussContent')
            ->orderBy(('discuss.created_at ASC'))
            ->limit(10)
            ->offset($offset)
            ->all();
        foreach ($table as $key => $item) {
            $result[] = array(
                'id' => intval($item['id']),
                'title' => $item->discussContent['name'],
                'content' => $item->discussContent['content'],
                'author' => $this->getAuthor($item['user_id']),
                'tags' => $item->discussContent['tags'],
                'mentions' => $item->discussContent['mentions'],
                'refer' => $item->discussContent['refer']
            );
        }
        return isset($result) ? $result : null;
```

Now when I have 1 result it works. When I have two it displays null in every line that should be filled by the mongo model. What can be the problem?

**Comment by klimov-paul** [comment]
What does `var_dump($item->discussContent)` shows?

**Comment by bdart** [comment]
`NULL` as well as when I `var_dump()` `$item` it shows that `discussContent` is empty. When there is just one item, it returns like expected:
```php
object(pirateship\versions\v1\models\MongoDiscussion)#282 (8) {
          ["_attributes":"yii\db\BaseActiveRecord":private]=>
          array(8) {
            ["_id"]=>
            object(MongoId)#283 (1) {
              ["$id"]=>
              string(24) "5538d5db459eb981430041a8"
            }
            ["title"]=>
            NULL
            ["content"]=>
            string(27) "hier ist neuer content wuhu"
            ["author"]=>
            int(46)
            ["tags"]=>
            NULL
            ["mentions"]=>
            NULL
            ["refer"]=>
            array(1) {
              [0]=>
              array(3) {
                ["id"]=>
                int(7)
                ["teaser"]=>
                string(30) "hier ist neuer content wuhu..."
                ["type"]=>
                string(10) "discussion"
              }
            }
            ["created_at"]=>
            int(1429788123)
          }
          ["_oldAttributes":"yii\db\BaseActiveRecord":private]=>
          array(8) {
            ["_id"]=>
            object(MongoId)#283 (1) {
              ["$id"]=>
              string(24) "5538d5db459eb981430041a8"
            }
            ["title"]=>
            NULL
            ["content"]=>
            string(27) "hier ist neuer content wuhu"
            ["author"]=>
            int(46)
            ["tags"]=>
            NULL
            ["mentions"]=>
            NULL
            ["refer"]=>
            array(1) {
              [0]=>
              array(3) {
                ["id"]=>
                int(7)
                ["teaser"]=>
                string(30) "hier ist neuer content wuhu..."
                ["type"]=>
                string(10) "discussion"
              }
            }
            ["created_at"]=>
            int(1429788123)
          }
          ["_related":"yii\db\BaseActiveRecord":private]=>
          array(0) {
          }
          ["_errors":"yii\base\Model":private]=>
          NULL
          ["_validators":"yii\base\Model":private]=>
          NULL
          ["_scenario":"yii\base\Model":private]=>
          string(7) "default"
          ["_events":"yii\base\Component":private]=>
          array(1) {
            ["beforeInsert"]=>
            array(1) {
              [0]=>
              array(2) {
                [0]=>
                array(2) {
                  [0]=>
                  object(yii\behaviors\TimestampBehavior)#281 (5) {
                    ["createdAtAttribute"]=>
                    string(10) "created_at"
                    ["updatedAtAttribute"]=>
                    string(10) "updated_at"
                    ["value"]=>
                    NULL
                    ["attributes"]=>
                    array(1) {
                      ["beforeInsert"]=>
                      array(1) {
                        [0]=>
                        string(10) "created_at"
                      }
                    }
                    ["owner"]=>
                    *RECURSION*
                  }
                  [1]=>
                  string(18) "evaluateAttributes"
                }
                [1]=>
                NULL
              }
            }
          }
          ["_behaviors":"yii\base\Component":private]=>
          array(1) {
            [0]=>
            object(yii\behaviors\TimestampBehavior)#281 (5) {
              ["createdAtAttribute"]=>
              string(10) "created_at"
              ["updatedAtAttribute"]=>
              string(10) "updated_at"
              ["value"]=>
              NULL
              ["attributes"]=>
              array(1) {
                ["beforeInsert"]=>
                array(1) {
                  [0]=>
                  string(10) "created_at"
                }
              }
              ["owner"]=>
              *RECURSION*
            }
          }
        }
```

**Comment by bdart** [comment]
I walked through the app.log and maybe it's up to the query?:
```php
2015-04-24 08:45:20 [::1][-][-][profile begin][yii\mongodb\Query::fetchRows]
find({
    "ns": "hangmanstree.discussion",
    "limit": 0,
    "batchSize": 0,
    "skip": 0,
    "flags": 0,
    "query": {
        "_id": {
            "$in": [
               { "$id": "5538d156459eb9b94e0041a8" },
               { "$id": "5538e4de459eb9e3010041a7" },
               { "$id": "5538eb27459eb9df010041a7" }
            ]
        }
    },
    "fields": [],
    "started_iterating": false
})
```
And this is when there is just one:
```php
2015-04-24 08:45:20 [::1][-][-][profile begin][yii\mongodb\Query::fetchRows]
find({
    "ns": "hangmanstree.discussion",
    "limit": 0,
    "batchSize": 0,
    "skip": 0,
    "flags": 0,
    "query": {
        "_id": {
            "$in": [
               { "$id": "5538d5db459eb981430041a8" }
            ]
        }
    },
    "fields": [],
    "started_iterating": false
})
```
Is this related to the cursor problem mentioned here some time ago? #6506 
```php
model.find({
    '_id': { $in: [
        '4ed3ede8844f0f351100000c',
        '4ed3f117a844e0471100000d', 
        '4ed3f18132f50c491100000e'
    ]}
}
```

**Comment by klimov-paul** [comment]
Can you provide exact values stored at 'discuss.storage_id'?

**Comment by bdart** [comment]
what do you mean by this? When i search for each item by itself it works. So the system finds the `storage_id` of each when I call a just `->one()`. Or do you mean something different?

**Comment by klimov-paul** [comment]
What does `var_dump($discuss->storage_id)` outputs?

**Comment by bdart** [comment]
`["storage_id"]=>string(24) "5538d156459eb9b94e0041a8"`

**Comment by klimov-paul** [comment]
What PHP Mongo extensio do you use?

**Comment by bdart** [comment]
composer.json:
`"yiisoft/yii2-mongodb": "~2.0",`

php info:
```
mongo

MongoDB Support => enabled
Version => 1.6.6
Streams Support => enabled
SSL Support => enabled
                   Supported Authentication Mechanisms
MONGODB-CR => enabled
SCRAM-SHA-1 => enabled
MONGODB-X509 => enabled
GSSAPI (Kerberos) => disabled
PLAIN => disabled

Directive => Local Value => Master Value
mongo.allow_empty_keys => 0 => 0
mongo.chunk_size => 261120 => 261120
mongo.cmd => $ => $
mongo.default_host => localhost => localhost
mongo.default_port => 27017 => 27017
mongo.is_master_interval => 15 => 15
mongo.long_as_object => 0 => 0
mongo.native_long => 1 => 1
mongo.ping_interval => 5 => 5
```

**Comment by klimov-paul** [comment]
This is odd.
Actually I am unable to reproduce the problem as you describe it.
To be more specific: in your case no 'MongoDiscussion' rows appar at all.
There is a problem at `ActiveRelationTrait::populateRelation()`, which prevents matching of plain string to `MongoId` instance.
I can fix it, but the question remains - how you even get any record at your relation?

**Comment by klimov-paul** [comment]
Issue resolved by commit c0fc459
